cpp/test/sg/fil_test.cu

```diff
@@ -421,16 +544,15 @@ class BaseFilTest : public testing::TestWithParam<FilTestParams> {
       want_preds_d, preds_d, ps.num_rows, raft::CompareApprox<float>(tolerance), stream));
   }
 
-  fil::val_t infer_one_tree(fil::dense_node* root, float* data)
+  fil::val_t infer_one_tree(fil::dense_node* root, float* data, const tree_base& tree)
   {
     int curr = 0;
     fil::val_t output{.f = 0.0f};
     for (;;) {
       const fil::dense_node& node = root[curr];
-      if (node.is_leaf()) return node.base_node::output<val_t>();
+      if (node.is_leaf()) return node.template output<val_t>();
       float val = data[node.fid()];
-      bool cond = isnan(val) ? !node.def_left() : val >= node.thresh();
-      curr      = (curr << 1) + 1 + (cond ? 1 : 0);
+      curr      = tree.get_child<true>(node, curr, val);
```

Comment:
This effectively compares the results obtained by calling `get_child` on host and on device. However, whether `get_child` returns the correct result isn't tested anywhere.

For testing, it's definitely better not to call `get_child` here, but to write the code that computes the index of the correct child node. In this way, the test would compare the results of two different implementations.